controller/model/api_session_heartbeats.go

```diff
@@ -0,0 +1,151 @@
+/*
+	Copyright NetFoundry, Inc.
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package model
+
+import (
+	"github.com/michaelquigley/pfxlog"
+	"github.com/openziti/edge/controller/persistence"
+	"github.com/openziti/foundation/storage/boltz"
+	"github.com/openziti/foundation/util/concurrenz"
+	cmap "github.com/orcaman/concurrent-map"
+	"time"
+)
+
+type HeartbeatCollector struct {
+	apiSessionLastUpdate cmap.ConcurrentMap //map[apiSessionId string] => *HeartbeatStatus
+	updateInterval       time.Duration
+	closeNotify          chan struct{}
+	isFlushing           concurrenz.AtomicBoolean
+	flushAction          func(beats []*Heartbeat)
+	batchSize            int
+}
+
+type Heartbeat struct {
+	ApiSessionId   string
+	LastActivityAt time.Time
+}
+
+// Creates a new HeartbeatCollector which is used to manage situations where an SDK is connecting to multiiple
+// Edge Routers and making API calls that all update their last updated at and trigger a write. The heartbeat
+// collector aggregates all of those calls into a single write and acts as an in memory buffer for last update
+// times.
+func NewHeartbeatCollector(env Env, batchSize int, updateInterval time.Duration, action func([]*Heartbeat)) *HeartbeatCollector {
+	collector := &HeartbeatCollector{
+		apiSessionLastUpdate: cmap.New(),
+		updateInterval:       updateInterval,
+		batchSize:            batchSize,
+		flushAction:          action,
+		closeNotify:          make(chan struct{}, 0),
+	}
+
```

Comment:
Is it worth adding this? We have to handle failed updates either way.

Reply:
Is it worth adding what?

Reply:
the api session event listener to remove entries from the map when session are removed. They'll get removed as they get flushed and we have to deal with write failures either way.

Reply:
Right now flushing doesn't remove them from the map. So the map can just grow.

Reply:
ah, ok